sanic/blueprint_group.py

```diff
@@ -0,0 +1,93 @@
+class BlueprintGroup(object):
+    """
+    This class provides a mechanism to implement a Blueprint Group
+    using the `Blueprint.group` method. To avoid having to re-write
+    some of the existing implementation, this class provides a custom
+    iterator implementation that will let you use the object of this
+    class as a list/tuple inside the existing implementation.
+    """
+
+    def __init__(self, url_prefix=None):
+        """
+        Create a new Blueprint Group
+
+        :param url_prefix: URL: to be prefixed before all the Blueprint Prefix
+        """
+        self._blueprints = list()
+        self._iter_position = 0
+        self._url_prefix = url_prefix
+
+    @property
+    def url_prefix(self):
+        """
+        Retrieve the URL prefix being used for the Current Blueprint Group
+        :return: string with url prefix
+        """
+        return self._url_prefix
+
+    @property
+    def blueprints(self):
+        """
+        Retrieve a list of all the available blueprints under this group.
+        :return: List of Blueprint instance
+        """
+        return self._blueprints
+
+    @blueprints.setter
+    def blueprints(self, blueprint):
+        """
+        Add a new Blueprint to the Group under consideration.
+        :param blueprint: Instance of Blueprint
+        :return: None
+        """
+        self._blueprints.append(blueprint)
+
+    def __iter__(self):
+        """Tun the class Blueprint Group into an Iterable item"""
+        return self
+
+    def __next__(self):
```

Comment:
It's basically iterating self._blueprints list here...I'd prefer
```
def __iter__(self):
    return iter(self. _blueprints)

# No need define __next__
```
And in current implementation, `_iter_position` always increments ? it can only iterate once ?

Reply:
@yunstanford Makes sense. The lesser code is better.

Reply:
@yunstanford But i think it's still a good idea to keep `__next__` because, it can help if you are using `next(x)` constructs to handle the iteratable item in different stages.

Reply:
As for `next()`, you could also utilize sth. like
```
bp_group_iter = iter(bp_group)
bp = next(bp_group_iter)
```

Reply:
I don't wanna make the class itself a iterator, which doesn't make sense.

But the class could provide methods like `.iterator()` or use `iter()`, that returns iterator.

Reply:
It'll have problems if you make the class itself iterator.

e.g. in this implementation, you maintain the iterator position `_iter_position`, and there is no way for two coroutines to iterate it at the same time, since they'll always have the same position.

also note, `_iter_position` always increments ? it can only iterate once ?

Reply:
@yunstanford As per the specs in PEP, isn't it expected that you get a `StopIteration` exception on exhaustion of the items in case of using `next()`?

> Iterator objects returned by either form of iter() have a next()
    method.  This method either returns the next value in the
    iteration, or raises StopIteration (or a derived exception class)
    to signal the end of the iteration.

The `__iter__` method has been modified to return `iter(self._blueprints)` so you can iterate over it multiple times in multiple places like using them in list comprehension etc. Am I missing something ?

Reply:
It'll raise `StopIteration` when exhausted all items. `iter([])` basically returns an iterator.

Reply:
@yunstanford That's what is being done now in the code right? The `__iter__` method supports the re-useable mechanism to iterate over the class object directly while `__next__` provides another mode while raising a `StopIteration`

Reply:
I don't see any benefit around adding current `__next__`, and it's quite confusing since it only supports to iterate once and after one full iteration, `_iter_position` is totally useless...

Reply:
@yunstanford No worries. Let me get rid of it.